### PR TITLE
Add support for the debug modules request

### PR DESCRIPTION
**Pull request by echarles**
This PR adds the support for the Modules request https://microsoft.github.io/debug-adapter-protocol/specification#Requests_Modules as discussed on https://github.com/jupyterlab/jupyterlab/pull/11566#discussion_r759661311

### CODE REVIEW

File: ipykernel/debugger.py [hunk 1]
```diff
@@ -562,6 +564,24 @@ async def richInspectVariables(self, message):
...
+            m = str(module)
+            if ".py'" in m:
+                a = {}
+                x = re.findall(r"'(.*?)'", m)
+                a[x[0]] = x[1]
+                mods.append(a)
+        reply = {
+            'body': {
+                'modules': mods,
+                'totalModules': len(modules)
```
Comment (JohanMabille) [inline ipykernel/debugger.py:580]: I think we the debugger should handle the `startModule` and `moduleCount` arguments speficied in the [DAP](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_Modules) and filter the result accordingly.

**Review by JohanMabille — CHANGES_REQUESTED** [review]

File: ipykernel/debugger.py [hunk 2]
```diff
@@ -562,6 +564,24 @@ async def richInspectVariables(self, message):
         reply["success"] = True
         return reply
 
+    async def modules(self, message):        
+        modules = [sys.modules[name] for name in sys.modules]
```
Comment (Carreau) [inline ipykernel/debugger.py:568]: Any reasons not to use `sys.modules.values()` here ?

File: ipykernel/debugger.py [hunk 3]
```diff
@@ -562,6 +564,24 @@ async def richInspectVariables(self, message):
...
+        mods = []
+        for module in modules:
+            m = str(module)
```
Comment (Carreau) [inline ipykernel/debugger.py:571]: Are you looking for `mod.__spec__.origin`  ? (`__spec__` may be non so you would need `getattr(..., '__spec__', None)`, so not sure it is easier.

**Review by Carreau — APPROVED** [review]
No objections from me.

File: ipykernel/debugger.py (hunk 3 above)
Comment (Carreau) [reply ipykernel/debugger.py:571]: I came  up with the following which is a bit less regexp-y:

```python
    async def modules(self, message):        
        modules = sys.modules.values()
        mods = []
        for module in modules:
            filename = getattr(getattr(module, "__spec__", None), "origin", None)
            if filename and filename.endswith(".py"):
                mods.append({mod.__name__: filename})

        reply = {"body": {"modules": mods, "totalModules": len(modules)}}
        return reply
```

One thing which is not clear to me is why would `mod.__name__` be different than the key in `sys.modules`, and is that roughly trying to do:

```
mods = [{k:m.__spec__.origin} for k,m in sys.modules.items()]
```

File: ipykernel/debugger.py [hunk 4]
```diff
@@ -263,7 +264,8 @@ class Debugger:
 
     # Requests that can be handled even if the debugger is not running
     static_debug_msg_types = [
-        'debugInfo', 'inspectVariables', 'richInspectVariables'
+        'debugInfo', 'inspectVariables', 
+        'richInspectVariables', 'modules'
     ]
```
[the remark this answers is not in the document]
Comment (echarles) [reply ipykernel/debugger.py:270]: Tracked in https://github.com/ipython/ipykernel/issues/833

File: ipykernel/debugger.py (hunk 2 above)
Comment (echarles) [reply ipykernel/debugger.py:587]: sys.modules.values() works. thx.

File: ipykernel/debugger.py (hunk 3 above)
Comment (echarles) [reply ipykernel/debugger.py]: I have implemented your suggestion in 8777e68e56712fef254ea72c70dca147b6e56a6d

Not sure how `sys.modules` should behave. I came to that code block in an empirical way.

File: ipykernel/debugger.py (hunk 1 above)
Comment (echarles) [reply ipykernel/debugger.py]: I have implemented the `startModule` and `moduleCount` in 95711b80b47537927bb2359daf43ea39cf2f0ad2

**Comment by echarles** [comment]
Current comments have been addressed.
I have rebased on `main` and undrafted.
CI is green.

This is ready for review/merge.

**Comment by JohanMabille** [comment]
LGTM

**Comment by echarles** [comment]
Thx @JohanMabille and @Carreau for the review and merge.

**Pull request merged**